pkg/skaffold/build/gcb/status.go

```diff
@@ -0,0 +1,272 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/api/cloudbuild/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/gcp"
+)
+
+var (
+	// maintain single instance of `statusReporter` per skaffold process
+	reporter     *statusManagerImpl
+	reporterOnce sync.Once
+	// maintain single instance of the GCB client per skaffold process
+	client     *cloudbuild.Service
+	clientOnce sync.Once
```

Comment:
Singletons are bad — they're a global dependency and breaks modularity.  We should strive to avoid them (look at the kubeContext/kubeConfig mess).  At the very least, have this instance be managed by the  `statusManagerImpl`.

You should be able to write tests for the statusManagerImpl and have different backoff values (nobody wants the test to wait a minute to verify timeout, for example).